### ISSUE TITLE
Formalize component health to include statuses and runtime health

### ISSUE THREAD

**Issue by binarylogic** [vectordotdev/vector#10555]
Operating Vector would benefit from a formalized concept of component health that could be collated in the `/health` endpoint. Right now the only concept of component health is the health check that is run at boot time, but there is no concept of runtime health, or classes of health.

For example, some problems should indicate that Vector has entered a non-recoverable failure state and should be rotated, while others should indicate degraded service but should not be rotated.

## Proposal

1. Components should report an ongoing health status. This is separate from, and in addition to, the health checks.
2. The health should be represented as a simple `red`, `yellow`, `green` status.
   * `red` - The component has entered a nonrecoverable failure state and should be rotated (i.e., a disk became unavailable for components that require a disk)
   * `yellow` - The component has entered a degraded failure state and should not be rotated (i.e., a downstream service is unreachable but may recover int he future) 
   * `green` - Everything is operating normally

### Disk buffers

* `red` - Disk is unavailable
* `yellow` - Disk is available but full, back pressure will be applied
* `green` - Disk is writable

### Elasticsearch sink

* `red` - Unable to open new connections due to a local system error (i.e., "too many open file handles")
* `yellow` - The downstream Elasticsearch cluster is unavailable
* `green` - The downstream Elasticsearch cluster is accepting data normally

**Comment by jszwedko** [comment]
> Vector has entered a non-recoverable failure state

Currently we just crash in these circumstances which I this is an alternative to the `red` state described here.

**Comment by spencergilbert** [comment]
@jszwedko @binarylogic to confirm, the consensus we came to was `red` statuses should be reserved for shutting down from failures that Vector deems irrecoverable (per @jszwedko's comment above around crashing as a `red` alternative). The response should indicate that Vector is shutting down, which component(s) triggered this, as well as why.

**Comment by binarylogic** [comment]
Agree, I'm going to adjust or close this issue. Given that component should just crash, there really is no such thing as a red status, so the classifications might be overkill. We should still open separate issues to properly represent startup and shutdown behavior in the health endpoint.

**Comment by spencergilbert** [comment]
> Agree, I'm going to adjust or close this issue. Given that component should just crash, there really is no such thing as a red status, so the classifications might be overkill. We should still open separate issues to properly represent startup and shutdown behavior in the health endpoint.

I think green/yellow are valid or some `healthy`/`degraded` separation. Red does seem unnecessary unless we can find a case where we can't function, but aren't sure restarting is the best option - perhaps some edge cases where we have data but may lose it on shutdown?
